File: tasks/native-runtime/fetch_url_parity_probe.py
"""fetch_url parity probe, Python side.

Pins `resolve_public_url`, the DNS-time SSRF check, HTML extraction, the cache,
redirect revalidation and the success envelope of `fetch_url` against
`deepseek_policy::fetch_url`.

DNS and HTTP are stubbed the same way `tests/test_tools.py` stubs them: a public
host resolves to `93.184.216.34`, and `public_http_connection` returns a scripted
response. The probe never opens a real socket.

Usage::

    python tasks/native-runtime/fetch_url_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example fetch_url_parity_probe > ../rust.json
    diff <(tr -d '\\r' < python.json) <(tr -d '\\r' < rust.json)
"""

from __future__ import annotations

import json
import os
import sys
import tempfile
from pathlib import Path
from typing import Any
from unittest.mock import patch

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

ROOT = tempfile.mkdtemp(prefix="fetch-url-parity-")
os.environ["DEEPSEEK_INFRA_ROOT"] = ROOT

from deepseek_infra.core.errors import AppError  # noqa: E402
from deepseek_infra.infra.rag.files import extract_html_text  # noqa: E402
from deepseek_infra.infra.tool_runtime import tools  # noqa: E402

PUBLIC_IP = "93.184.216.34"
NOW = 1_700_000_000.0
HTML = (
    b"<html><head><title>Doc</title><script>secret()</script></head>"
    b"<body><main><h1>Hello</h1><p>Readable page text.</p></main></body></html>"
)

RESOLVE_CASES = [
    ("https-path-query-fragment", "https://example.com/path?x=1#fragment"),
    ("explicit-default-port", "https://example.com:443/x"),
    ("http-default-port", "http://example.com/x"),
    ("explicit-http-port", "http://example.com:8080/x"),
    ("loopback", "http://127.0.0.1:8000/"),
    ("localhost", "http://localhost:8000"),
    ("local-suffix", "http://printer.local/"),
    ("ftp", "ftp://example.com"),
    ("credentials", "https://user:pass@example.com"),
    ("empty", ""),
    ("blank", "   "),
    ("no-host", "https://"),
    ("bad-port", "https://example.com:abc/"),
    ("metadata-ip", "http://169.254.169.254/"),
]


_real_resolve = tools.resolve_public_host


def resolve_host(host: str, port: int | None) -> list[str]:
    if host == "example.com":
        return [PUBLIC_IP]
    return _real_resolve(host, port)


def error_view(exc: AppError) -> dict[str, Any]:
    return {"error": str(exc), "code": exc.code.value, "status": exc.status}


def target_view(target: tools.PublicUrlTarget) -> dict[str, Any]:
    return {
        "url": target.url,
        "scheme": target.scheme,
        "host": target.host,
        "port": target.port,
        "hostHeader": target.host_header,
        "requestTarget": target.request_target,
        "address": target.address,
    }


class FakeResponse:
    def __init__(self, data: bytes, *, status: int = 200, headers: dict[str, str] | None = None) -> None:
        self.data = data
        self.status = status
        self.headers = {"Content-Type": "text/html; charset=utf-8", **(headers or {})}

    def getheader(self, name: str, default: str = "") -> str:
        return self.headers.get(name, default)

    def read(self, size: int = -1) -> bytes:
        if size >= 0:
            return self.data[:size]
        return self.data


class FakeConnection:
    def __init__(self, response: FakeResponse) -> None:
        self.response = response
        self.target: tools.PublicUrlTarget | None = None

    def request(self, *args: object, **kwargs: object) -> None:
        return None

    def getresponse(self) -> FakeResponse:
        return self.response

    def close(self) -> None:
        return None


def main() -> int:
    out: dict[str, Any] = {}

    with patch.object(tools, "resolve_public_host", side_effect=resolve_host):
        for label, url in RESOLVE_CASES:
            try:
                out[f"resolve::{label}"] = target_view(tools.resolve_public_url(url))
            except AppError as exc:
                out[f"resolve::{label}"] = error_view(exc)

        tools.ensure_public_address(PUBLIC_IP)
        out["address::public"] = "ok"
        for label, address in [
            ("loopback", "127.0.0.1"),
            ("metadata", "169.254.169.254"),
            ("cgnat", "100.64.0.1"),
            ("mapped-loopback", "::ffff:127.0.0.1"),
        ]:
            try:
                tools.ensure_public_address(address)
                out[f"address::{label}"] = "ok"
            except AppError as exc:
                out[f"address::{label}"] = error_view(exc)

        out["extract::html"] = extract_html_text(HTML)
        out["extract::plain"] = tools.extract_readable_text(b"just text\n\n\nmore", "text/plain")
        out["normalize"] = tools.normalize_text("  a   b\n\n\n\nc  ")

        connections: list[FakeConnection] = []

        def fake_ok(target: tools.PublicUrlTarget, timeout: float) -> FakeConnection:
            connection = FakeConnection(FakeResponse(HTML))
            connection.target = target
            connections.append(connection)
            return connection

        with patch.object(tools, "time") as time_module:
            time_module.time.return_value = NOW
            with patch.object(tools, "public_http_connection", side_effect=fake_ok):
                first = tools.fetch_url("https://example.com/path?x=1#fragment")
                second = tools.fetch_url("https://example.com/path?x=1#fragment")
        out["fetch::first"] = first
        out["fetch::second"] = second
        out["fetch::http-hits"] = len(connections)
        if connections:
            out["fetch::pinned-address"] = connections[0].target.address if connections[0].target else None
            out["fetch::host-header"] = connections[0].target.host_header if connections[0].target else None

        def fake_redirect(target: tools.PublicUrlTarget, timeout: float) -> FakeConnection:
            return FakeConnection(FakeResponse(b"", status=302, headers={"Location": "http://127.0.0.1/admin"}))

        with patch.object(tools, "public_http_connection", side_effect=fake_redirect):
            try:
                tools.fetch_url("https://example.com/path")
                out["fetch::redirect"] = "ok"
            except AppError as exc:
                out["fetch::redirect"] = error_view(exc)

        huge = b"x" * (tools.MAX_FETCH_BYTES + 10)

        def fake_huge(target: tools.PublicUrlTarget, timeout: float) -> FakeConnection:
            return FakeConnection(FakeResponse(huge, headers={"Content-Type": "text/html"}))

        with patch.object(tools, "public_http_connection", side_effect=fake_huge):
            try:
                tools.fetch_url("https://example.com/huge")
                out["fetch::too-large"] = "ok"
            except AppError as exc:
                out["fetch::too-large"] = error_view(exc)

        def fake_503(target: tools.PublicUrlTarget, timeout: float) -> FakeConnection:
            return FakeConnection(FakeResponse(b"", status=503))

        with patch.object(tools, "public_http_connection", side_effect=fake_503):
            try:
                tools.fetch_url("https://example.com/down")
                out["fetch::http-503"] = "ok"
            except AppError as exc:
                out["fetch::http-503"] = error_view(exc)

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/fetch_url_parity_probe.rs
//! fetch_url parity probe, Rust side.
//!
//! Replays the same scripted corpus as
//! `tasks/native-runtime/fetch_url_parity_probe.py` through
//! `deepseek_policy::fetch_url` and prints canonical JSON.
//!
//! Usage::
//!
//!     python tasks/native-runtime/fetch_url_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example fetch_url_parity_probe > ../rust.json
//!     diff <(tr -d '\r' < python.json) <(tr -d '\r' < rust.json)

use std::cell::RefCell;
use std::fs;
use std::path::PathBuf;

use deepseek_policy::app_error::AppError;
use deepseek_policy::fetch_url::{
    FetchContext, HttpResponse, MAX_FETCH_BYTES, PublicUrlTarget, ensure_public_address,
    extract_html_text, extract_readable_text, fetch_url, normalize_text, resolve_public_url,
};
use serde_json::{Map, Value, json};

const PUBLIC_IP: &str = "93.184.216.34";
const NOW: f64 = 1_700_000_000.0;
const HTML: &[u8] = b"<html><head><title>Doc</title><script>secret()</script></head><body><main><h1>Hello</h1><p>Readable page text.</p></main></body></html>";

fn public_dns(host: &str, _port: Option<u16>) -> Result<Vec<String>, AppError> {
    if host == "example.com" {
        return Ok(vec![PUBLIC_IP.to_string()]);
    }
    // Names the probe never asks this stub to resolve. A literal IP is handled
    // inside `resolve_public_url` before DNS is consulted.
    Err(AppError {
        message: "URL host could not be resolved".to_string(),
        code: deepseek_policy::app_error::codes::UPSTREAM_FAILURE,
        status: 502,
    })
}

fn error_view(error: &AppError) -> Value {
    json!({"error": error.message, "code": error.code, "status": error.status})
}

fn target_view(target: &PublicUrlTarget) -> Value {
    json!({
        "url": target.url,
        "scheme": target.scheme,
        "host": target.host,
        "port": target.port,
        "hostHeader": target.host_header,
        "requestTarget": target.request_target,
        "address": target.address,
    })
}

fn resolve_result(url: &str) -> Value {
    match resolve_public_url(url, &public_dns) {
        Ok(target) => target_view(&target),
        Err(error) => error_view(&error),
    }
}

fn address_result(address: &str) -> Value {
    match ensure_public_address(address) {
        Ok(()) => json!("ok"),
        Err(error) => error_view(&error),
    }
}

fn fetch_result(
    url: &str,
    http: &deepseek_policy::fetch_url::HttpGet<'_>,
    root: &std::path::Path,
) -> Value {
    let context = FetchContext {
        root,
        now_epoch: NOW,
        timeout_seconds: 45,
        dns: &public_dns,
        http,
    };
    match fetch_url(url, Some(&context)) {
        Ok(value) => value,
        Err(error) => error_view(&error),
    }
}

fn main() {
    let mut out = Map::new();
    let resolve_cases = [
        (
            "https-path-query-fragment",
            "https://example.com/path?x=1#fragment",
        ),
        ("explicit-default-port", "https://example.com:443/x"),
        ("http-default-port", "http://example.com/x"),
        ("explicit-http-port", "http://example.com:8080/x"),
        ("loopback", "http://127.0.0.1:8000/"),
        ("localhost", "http://localhost:8000"),
        ("local-suffix", "http://printer.local/"),
        ("ftp", "ftp://example.com"),
        ("credentials", "https://user:pass@example.com"),
        ("empty", ""),
        ("blank", "   "),
        ("no-host", "https://"),
        ("bad-port", "https://example.com:abc/"),
        ("metadata-ip", "http://169.254.169.254/"),
    ];
    for (label, url) in resolve_cases {
        out.insert(format!("resolve::{label}"), resolve_result(url));
    }

    out.insert("address::public".to_string(), address_result(PUBLIC_IP));
    out.insert("address::loopback".to_string(), address_result("127.0.0.1"));
    out.insert(
        "address::metadata".to_string(),
        address_result("169.254.169.254"),
    );
    out.insert("address::cgnat".to_string(), address_result("100.64.0.1"));
    out.insert(
        "address::mapped-loopback".to_string(),
        address_result("::ffff:127.0.0.1"),
    );

    out.insert("extract::html".to_string(), json!(extract_html_text(HTML)));
    out.insert(
        "extract::plain".to_string(),
        json!(extract_readable_text(b"just text\n\n\nmore", "text/plain")),
    );
    out.insert(
        "normalize".to_string(),
        json!(normalize_text("  a   b\n\n\n\nc  ")),
    );

    let root: PathBuf =
        std::env::temp_dir().join(format!("fetch-url-parity-{}", std::process::id()));
    let _ = fs::remove_dir_all(&root);
    fs::create_dir_all(&root).expect("scratch");

    let hits = RefCell::new(0_u32);
    let last_target: RefCell<Option<PublicUrlTarget>> = RefCell::new(None);
    let ok_http = |target: &PublicUrlTarget, _timeout: u64| {
        *hits.borrow_mut() += 1;
        last_target.replace(Some(target.clone()));
        Ok(HttpResponse {
            status: 200,
            location: String::new(),
            content_type: "text/html; charset=utf-8".to_string(),
            body: HTML.to_vec(),
        })
    };
    out.insert(
        "fetch::first".to_string(),
        fetch_result("https://example.com/path?x=1#fragment", &ok_http, &root),
    );
    out.insert(
        "fetch::second".to_string(),
        fetch_result("https://example.com/path?x=1#fragment", &ok_http, &root),
    );
    out.insert("fetch::http-hits".to_string(), json!(*hits.borrow()));
    let pinned = last_target.borrow();
    out.insert(
        "fetch::pinned-address".to_string(),
        json!(pinned.as_ref().map(|target| target.address.clone())),
    );
    out.insert(
        "fetch::host-header".to_string(),
        json!(pinned.as_ref().map(|target| target.host_header.clone())),
    );

    let redirect_http = |_target: &PublicUrlTarget, _timeout: u64| {
        Ok(HttpResponse {
            status: 302,
            location: "http://127.0.0.1/admin".to_string(),
            content_type: String::new(),
            body: Vec::new(),
        })
    };
    out.insert(
        "fetch::redirect".to_string(),
        fetch_result("https://example.com/path", &redirect_http, &root),
    );

    let huge = vec![b'x'; MAX_FETCH_BYTES + 10];
    let huge_http = |_target: &PublicUrlTarget, _timeout: u64| {
        Ok(HttpResponse {
            status: 200,
            location: String::new(),
            content_type: "text/html".to_string(),
            body: huge.clone(),
        })
    };
    out.insert(
        "fetch::too-large".to_string(),
        fetch_result("https://example.com/huge", &huge_http, &root),
    );

    let down_http = |_target: &PublicUrlTarget, _timeout: u64| {
        Ok(HttpResponse {
            status: 503,
            location: String::new(),
            content_type: String::new(),
            body: Vec::new(),
        })
    };
    out.insert(
        "fetch::http-503".to_string(),
        fetch_result("https://example.com/down", &down_http, &root),
    );

    let _ = fs::remove_dir_all(&root);
    let mut encoded = serde_json::to_string_pretty(&Value::Object(out)).expect("serialize");
    encoded.push('\n');
    print!("{encoded}");
}
